include/core/CContainerPrinter.h

```diff
@@ -223,172 +198,211 @@ class CNodePrinter<true_, PRINTER> {
     }
 
     //! Print a std::pair.
     template<typename U, typename V>
     static std::string printElement(const std::pair<U, V>& value) {
-        std::ostringstream result;
-        result << "(" << printElement(ml::core::unwrap_ref(value.first)) << ", "
-               << printElement(ml::core::unwrap_ref(value.second)) << ")";
-        return result.str();
+        std::string result;
+        result.reserve(2 * (SIZE_OF_ELEMENT + 1) + 2);
+        result += "(";
+        result += printElement(value.first);
+        result += ", ";
+        result += printElement(value.second);
+        result += ")";
+        return result;
     }
 
     //! Print a std::tuple.
     template<typename... ARGS>
     static std::string printElement(const std::tuple<ARGS...>& value) {
-        std::ostringstream result;
+        std::string result;
+        result.reserve((SIZE_OF_ELEMENT + 1) * sizeof...(ARGS) + 2);
         std::apply(
             [&result](ARGS const&... args) {
-                result << '(';
+                result += '(';
                 std::size_t n{0};
-                ((result << args << (++n != sizeof...(ARGS) ? ", " : "")), ...);
-                result << ')';
+                ((result += printElement(args) + (++n != sizeof...(ARGS) ? ", " : "")), ...);
+                result += ')';
             },
             value);
-        return result.str();
+        return result;
+    }
+
+    //! Print a string.
+    static const std::string& printElement(const std::string& value) {
+        return value;
     }
 
-    //! Print a string for debug (otherwise we split them into their
-    //! component characters since they have iterators).
-    static std::string printElement(const std::string& value) { return value; }
+    //! Print a CStoredStringPtr.
+    static const std::string& printElement(const CStoredStringPtr& value);
 
 public:
-    //! Function object wrapper around printElement for use with STL.
-    class CElementPrinter {
-    public:
-        template<typename T>
-        std::string operator()(const T& value) {
-            return printElement(value);
-        }
-    };
+    //! Fallback print.
+    template<typename T>
+    static auto print(const T& t) -> decltype(printElement(t)) {
+        return printElement(t);
+    }
 
-    //! Print a range of values as defined by a start and end iterator
-    //! for debug. This assumes that ITR is a forward iterator, i.e.
-    //! it must implement prefix ++ and * operators.
+    //! Print a range of values as defined by a start and end iterator. This
+    //! assumes that ITR is a forward iterator, i.e. it must implement prefix
+    //! ++ and * operators.
     template<typename ITR>
     static std::string print(ITR begin, ITR end) {
-        std::ostringstream result;
-
-        result << "[";
+        std::string result;
+        result.reserve((SIZE_OF_ELEMENT + 1) * std::distance(begin, end) + 2);
```

Comment:
Good point, it would be more consistent; I'll change that. Of course this is kind of a guess because the element itself could be a container. It would probably be worthwhile here actually adding in a function to compute size properly without actually printing if we're going for performance, but this isn't really designed for performance. I just went with something not totally unreasonable.